quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/KeycloakProcessor.java

```diff
@@ -272,7 +273,18 @@ private void configureDefaultPersistenceUnitProperties(ParsedPersistenceXmlDescr
         Properties unitProperties = descriptor.getProperties();
 
         unitProperties.setProperty(AvailableSettings.DIALECT, config.defaultPersistenceUnit.dialect.dialect.orElse(null));
-        unitProperties.setProperty(AvailableSettings.JPA_TRANSACTION_TYPE, PersistenceUnitTransactionType.JTA.name());
+        String storage = Configuration.getRawValue(
+                MicroProfileConfigProvider.NS_KEYCLOAK_PREFIX.concat(StorageOptions.STORAGE.getKey()));
+        if (storage == null) {
+            // legacy store
+            unitProperties.setProperty(AvailableSettings.JPA_TRANSACTION_TYPE, PersistenceUnitTransactionType.JTA.name());
+        } else {
+            // JPA map store; avoid JTA and use a direct (resource local) approach
+            unitProperties.setProperty(AvailableSettings.JPA_TRANSACTION_TYPE, PersistenceUnitTransactionType.RESOURCE_LOCAL.name());
+            // needed to change the connection handling to avoid Hibernate returning the connection too early,
+            // which then interfered with the auto-commit reset that's done at the end of the transaction and PgConnection throwing a "Cannot commit when autoCommit is enabled."
+            unitProperties.setProperty(AvailableSettings.CONNECTION_HANDLING, PhysicalConnectionHandlingMode.DELAYED_ACQUISITION_AND_RELEASE_AFTER_TRANSACTION.name());
```

Comment:
See https://github.com/quarkusio/quarkus/blob/main/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/boot/FastBootMetadataBuilder.java#L287.

In the past, we had some discussions with Quarkus team about it. We can involve them again to see the implications of doing this. At that time, the discussions were around JTA, perhaps using resource local it can work well.

Reply:
As agreed in today's call, I'll reach out to experts in Quarkus/Hibernate/Agroal to discuss this. Still, this PR will be merged to fix the problems we currently experience.